File: UniBlock/src/main/scala/data/event/Encryptable.scala
package it.unifi.nave.uniblock
package data.event

import data.Hashable
import data.event.Encryptable.EventType

import java.time.LocalDate

object Encryptable extends Enumeration {
  type EventType = Value
  val Genesis, Certificate, ExamPublishing, ExamBooking, ExamResult, ExamConfirm = Value
}

sealed trait Encryptable extends Hashable {
  def getType: EventType
}

case class ExamPublishing(professor: String, codeExam: String, date: LocalDate) extends Encryptable {
  override def getType: EventType = Encryptable.ExamPublishing
}

case class ExamBooking(student: String, codeExam: String, date: LocalDate) extends Encryptable {
  override def getType: EventType = Encryptable.ExamBooking
}

case class ExamResult(professor: String, student: String, codeExam: String, date: LocalDate, result: Int) extends Encryptable {
  override def getType: EventType = Encryptable.ExamResult
}

case class ExamConfirm(student: String, codeExam: String, date: LocalDate, confirm: Boolean) extends Encryptable {
  override def getType: EventType = Encryptable.ExamConfirm
}

File: program/src/main/java/it/unifi/nave/uniblock/data/event/Encryptable.java
package it.unifi.nave.uniblock.data.event;

import java.io.Serializable;
import java.time.LocalDate;

public interface Encryptable extends Serializable {

    EventType getType();

    enum EventType {
        EXAM_PUBLISHING, EXAM_BOOKING, EXAM_RESULT, EXAM_CONFIRM
    }

    record ExamPublishing(String professor, String codeExam, LocalDate date) implements Encryptable {
        @Override
        public EventType getType() {
            return EventType.EXAM_PUBLISHING;
        }
    }

    record ExamBooking(String student, String codeExam, LocalDate date) implements Encryptable {
        @Override
        public EventType getType() {
            return EventType.EXAM_BOOKING;
        }
    }

    record ExamResult(String professor, String student, String codeExam, LocalDate date,
                      int result) implements Encryptable {
        @Override
        public EventType getType() {
            return EventType.EXAM_RESULT;
        }
    }

    record ExamConfirm(String student, String codeExam, LocalDate date, boolean confirm) implements Encryptable {
        @Override
        public EventType getType() {
            return EventType.EXAM_CONFIRM;
        }
    }
}
